M2/Macaulay2/packages/Colon.m2

```diff
@@ -0,0 +1,885 @@
+---------------------------------------------------------------------------
+-- PURPOSE : Computation of quotient, saturation, and annihilator
+--
+-- UPDATE HISTORY : created 14 April 2018 at M2@UW;
+--                  updated November 2020
+--
+-- TODO : 1. why are there shadowed symbols?
+--        2. any strategies to incorporate from Shimoyama-Yokoyama.m2
+--           or greedySat in GTZ.m2 or getSaturation in newGTZ.m2
+--        3. which algorithms for Ideals can be adapted to submodules?
+--        4. move radical here, perhaps intersect as well
+--        5. employ storefun! Maybe call it declare
+---------------------------------------------------------------------------
+newPackage(
+    "Colon",
+    Version => "0.2",
+    Date => "November 8, 2020",
+    Headline => "quotient, saturation, and annihilator routines for ideals and modules",
+    Authors => {
+	{Name => "Justin Chen",    Email => "justin.chen@math.gatech.edu"},
+	{Name => "Mahrud Sayrafi", Email => "mahrud@umn.edu",        HomePage => "https://math.umn.edu/~mahrud"},
+	{Name => "Mike Stillman",  Email => "mike@math.cornell.edu", HomePage => "http://www.math.cornell.edu/~mike"}},
+    Keywords => {"Commutative Algebra"},
+    PackageExports => { "Elimination" },
+    AuxiliaryFiles => true,
+    DebuggingMode => false
+    )
+
+export { "isSupportedInZeroLocus" }
+
+exportFrom_Core { "saturate", "annihilator" }
+
+importFrom_Core { "printerr", "raw", "rawColon", "rawSaturate", "newMonomialIdeal", "eliminationInfo" }
+
+-- TODO: where should these be placed?
+trim MonomialIdeal := MonomialIdeal => opts -> (cacheValue (symbol trim => opts)) ((I) -> monomialIdeal trim(module I, opts))
+
+-- TODO: is this the right function?
+ambient Ideal := Ideal => I -> ideal 1_(ring I)
```

Comment:
If you look at the other place where I've used `ambient` in this code you'll see that I wanted a function that behaves similarly for ideals and modules. In particular, all ideals are rank 1 modules over the ring, and the unit ideal is canonically the same as R^1. So if you consider an ideal as a module, it's ambient is the unit ideal.

Reply:
It's still not similar, because modules are actually implemented in such a way that `ambient M` is referred to by M.

Reply:
> It's still not similar, because modules are actually implemented in such a way that `ambient M` is referred to by M.

What??
```m2
      ambient Module := Module => M -> (
           if M.?generators then M.generators.target
           else if M.?relations then M.relations.target
           else M)
```
The code I wrote is simply interpreting an ideal as a module with generators, and returns `target gens I`. How is that not similar?

Reply:
> > It's still not similar, because modules are actually implemented in such a way that `ambient M` is referred to by M.
> 
> What?? Are you sure you're talking about `ambient`?

Yes, look at this example:
```m2

i1 : R=QQ[x];

i2 : M = subquotient (matrix {{x}}, matrix {{x^2}})

o2 = subquotient (| x |, | x2 |)

                               1
o2 : R-module, subquotient of R

i6 : target M.relations === ambient M

o6 = true
```

> 
> ```m2
>       ambient Module := Module => M -> (
>            if M.?generators then M.generators.target
>            else if M.?relations then M.relations.target
>            else M)
> ```
> 
> The code I wrote is simply interpreting an ideal as a module with generators, and returns `target gens I`. How is that not similar?

No, your code doesn't return `target gens I` -- it returns the corresponding ideal, which is always the unit ideal.  Mathematically uninteresting and unuseful for software purposes.

Reply:
Obviously I found this both mathematically meaningful (as I explained) and useful for simplifying a software I wrote. Specifically, I use it to simplify the code in more than one place:
https://github.com/Macaulay2/M2/blob/a90d7f2aca8a974508569dd753e5845f2e8b40b7/M2/Macaulay2/packages/Colon.m2#L196
https://github.com/Macaulay2/M2/blob/a90d7f2aca8a974508569dd753e5845f2e8b40b7/M2/Macaulay2/packages/Colon.m2#L407
https://github.com/Macaulay2/M2/blob/a90d7f2aca8a974508569dd753e5845f2e8b40b7/M2/Macaulay2/packages/Colon.m2#L411
In all cases, defining ambient of an ideal to be the unit ideal means I don't have to add an extra condition based on the class of A. This makes the code easier to read for a mathematician as well, because you can simply consider A as a rank 1 module.

If this already exists or if there's a better name for a function which for modules gives the ambient and for ideals gives the unit ideal, please let me know. Otherwise, I'm confused what it is that you want here.

Reply:
If it helps resolve this issue and move on, I'll just change it to an internal function `ambient'`.

Reply:
Hmm, in this line:
```m2
if class B =!= Module and isSubset(ambient A, B) then return A; 
```
, apparently you are testing whether B is an ideal and is the unit ideal -- the latter condition has nothing
to do with A, so simpler code would be this:
```m2
if instance(B,Ideal) and B == 1 then return A;
```

As for adding extra conditions based on the class of A -- why not just define `I:J` to be `module I : module J`?  Then the code would always be dealing just with modules, and there would be no such tests.

Hmm, I see you've already switched it to `ambient'` and resolved the issue without waiting for a response from me....

Reply:
You're right that it could be further simplified, but I didn't just check B == 1 because a TODO item is to speed up computation of A:B for two modules if we know that A\subset B. I'm not sure what the answer is yet, but once I know, that's what will happen for the case where B is a module. I'm guessing it has to do with their primary decompositions.

> Hmm, I see you've already switched it to ambient' and resolved the issue without waiting for a response from me....

I'm eager to finish this PR and move on. We can talk about what ambient of an ideal is later.